### PR TITLE
add link status detection

### PR DESCRIPTION
**Pull request by zshi-redhat**
Detects Link status for Linux network devices and
updates associated VFs health state accordingly every
5 seconds.

Add GetVFList util and simplify discoverNetworks.

Add rootDevices in sriovManager to track discovered
and SR-IOV enabled PF devices.

### CODE REVIEW

**Comment by s1061123** [comment]
How about to use [netlink's LinkSubscribe](https://godoc.org/github.com/vishvananda/netlink#LinkSubscribe) for monitoring link status?
With this API, we don't have to polling interface and get events from kernel through netlink.

I understand that could be trade-off and there is no absolute 'best' solution, so it is just an idea/comment.

File: cmd/sriovdp/sriov-device-plugin.go [hunk 1]
```diff
@@ -149,41 +181,49 @@ func (sm *sriovManager) discoverNetworks() error {
...
-				if err != nil {
-					glog.Errorf("Error. Cannot read symbolic link between virtual function and PCI - Device: %s, VF: %v. Err: %v", dev, vf, err)
-					return err
-				}
-
-				pciAddr := pciInfo[len("../"):]
-				glog.Infof("PCI Address for device %s, VF %v is %s", dev, vf, pciAddr)
-
-				devName := pciAddr
-				sm.devices[devName] = pluginapi.Device{ID: devName, Health: pluginapi.Healthy}
```
Comment (ahalim-intel) [inline cmd/sriovdp/sriov-device-plugin.go]: I see you have refactored this section and extracted the GetVFList() method. But I don't see the initial device list is being populated anywhere. We only get a list of device only after a Probe() is invoked. The ListAndWatch() should send the the initial device list(which is now empty) first and will send updated list only when any device's health is changed.

File: cmd/sriovdp/sriov-device-plugin.go [hunk 2]
```diff
@@ -149,41 +181,49 @@ func (sm *sriovManager) discoverNetworks() error {
...
 		}
 	}
+	glog.Infof("Discovered SR-IOV PF devices: %v", sm.rootDevices)
 	return nil
 }
 
-func (sm *sriovManager) GetDeviceState(DeviceName string) string {
-	// TODO: Discover device health
-	return pluginapi.Healthy
+func IsNetlinkStatusUp(dev string) bool {
```
[the remark this answers is not in the document]
Comment (zshi-redhat) [reply cmd/sriovdp/sriov-device-plugin.go]: +1, will add comment.

File: cmd/sriovdp/sriov-device-plugin.go [hunk 3]
```diff
@@ -149,41 +181,49 @@ func (sm *sriovManager) discoverNetworks() error {
...
+	opsFile := filepath.Join(netDirectory, dev, "operstate")
+	bytes, err := ioutil.ReadFile(opsFile)
+	if err != nil || strings.TrimSpace(string(bytes)) != "up" {
+		return false
+	}
+	return true
+}
+
+func (sm *sriovManager) Probe() bool {
```
[the remark this answers is not in the document]
Comment (zshi-redhat) [reply cmd/sriovdp/sriov-device-plugin.go]: +1, will add comment.

File: cmd/sriovdp/sriov-device-plugin.go (hunk 1 above)
Comment (zshi-redhat) [reply cmd/sriovdp/sriov-device-plugin.go]: The ListAndWatch() updates device information by calling Probe() first; when sm.devices[] list is empty, Probe() assigns default value to it and set the 'changed' flag to true which triggers the ListAndWatch to send ListAndWatchResponse back. The initial assignment of sm.devices[] is at https://github.com/intel/sriov-network-device-plugin/pull/32/files#diff-8dbc0a94295122ac9cb2460632637cbdR220

Comment (ahalim-intel) [reply cmd/sriovdp/sriov-device-plugin.go]: Ok, thanks for explanation.

**Comment by ahalim-intel** [comment]
> How about to use [netlink's LinkSubscribe](https://godoc.org/github.com/vishvananda/netlink#LinkSubscribe) for monitoring link status?
> With this API, we don't have to polling interface and get events from kernel through netlink.
> 
> I understand that could be trade-off and there is no absolute 'best' solution, so it is just an idea/comment.

This might be a good solution going forward. But it will need a good bit of efforts to implementation. Maybe something we can look into in future as an improvement.

**Comment by zshi-redhat** [comment]
> > How about to use [netlink's LinkSubscribe](https://godoc.org/github.com/vishvananda/netlink#LinkSubscribe) for monitoring link status?
> > With this API, we don't have to polling interface and get events from kernel through netlink.
> > I understand that could be trade-off and there is no absolute 'best' solution, so it is just an idea/comment.
> 
> This might be a good solution going forward. But it will need a good bit of efforts to implementation. Maybe something we can look into in future as an improvement.

+1 on the notification feature for link state change and agree with Abdul that we could look into it in future, Implement it on master branch first, then backport to release.

**Comment by s1061123** [comment]
> > > How about to use [netlink's LinkSubscribe](https://godoc.org/github.com/vishvananda/netlink#LinkSubscribe) for monitoring link status?
> > > With this API, we don't have to polling interface and get events from kernel through netlink.
> > > I understand that could be trade-off and there is no absolute 'best' solution, so it is just an idea/comment.
> > 
> > 
> > This might be a good solution going forward. But it will need a good bit of efforts to implementation. Maybe something we can look into in future as an improvement.
> 
> +1 on the notification feature for link state change and agree with Abdul that we could look into it in future, Implement it on master branch first, then backport to release.

I've implement prototype with netlink LinkSubscribe(). It is not tested because I don't have any SR-IOV device ready testbed. Just FYI.
https://github.com/s1061123/sriov-network-device-plugin/tree/dev/prototype-netlink-status

**Comment by ahalim-intel** [comment]
@s1061123 @zshi-redhat 
I have added a comment here: 
https://github.com/s1061123/sriov-network-device-plugin/commit/55ae86fd7fcf627ece2631ab40d6fe333f291b25#commitcomment-31301361

**Pull request merged**